test/e2e-test/appcontext_test.go

```diff
@@ -0,0 +1,315 @@
+
+			ApplicationConfiguration: util.Object2RawExtension(ac1),
+			Application:              *dummyApp,
+		}}
+	trait2 := &v1alpha2.ManualScalerTrait{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha2.ManualScalerTraitKind,
+			APIVersion: v1alpha2.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      traitName2,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ManualScalerTraitSpec{
+			ReplicaCount: 3,
+		},
+	}
+	rawTrait2 := runtime.RawExtension{Object: trait2}
+	ac2 := ac(
+		acWithName(acName2),
+		acWithNamspace(namespace),
+		acWithComps([]v1alpha2.ApplicationConfigurationComponent{
+			{
+				ComponentName: compName2,
+				Traits: []v1alpha2.ComponentTrait{
+					v1alpha2.ComponentTrait{
+						Trait: rawTrait2,
+					},
+				},
+			},
+		}),
+	)
+	ar2 := &v1alpha2.ApplicationRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      arName2,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationRevisionSpec{
+			ApplicationConfiguration: util.Object2RawExtension(ac2),
+			Components:               convertComponentList2Map([]*v1alpha2.Component{co2}),
+			Application:              *dummyApp,
+		}}
+	appContext := &v1alpha2.ApplicationContext{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      appContextName,
+			Namespace: namespace,
+		},
+		Spec: v1alpha2.ApplicationContextSpec{
+			ApplicationRevisionName: arName1,
+		},
+	}
+
+	BeforeEach(func() {
+		Expect(k8sClient.Create(ctx, &ns)).Should(SatisfyAny(BeNil()))
+		Eventually(
+			func() error {
+				return k8sClient.Get(ctx, types.NamespacedName{Name: namespace}, &ns)
+			},
+			time.Second*3, time.Millisecond*300).Should(BeNil())
+
+		Expect(k8sClient.Create(ctx, co1)).Should(Succeed())
+		Expect(k8sClient.Create(ctx, co2)).Should(Succeed())
+		Expect(k8sClient.Create(ctx, ar1)).Should(Succeed())
+		Expect(k8sClient.Create(ctx, ar2)).Should(Succeed())
+	})
+
+	AfterEach(func() {
+		By("Clean up resources after a test")
+		Expect(k8sClient.Delete(ctx, &ns, client.PropagationPolicy(metav1.DeletePropagationForeground))).Should(Succeed())
+		time.Sleep(15 * time.Second)
+	})
+
+	It("Test appContext reconcil logic ", func() {
+		By("Test AppRevision1 only have 1 workload on trait")
+		Expect(k8sClient.Create(ctx, appContext)).Should(Succeed())
+		Eventually(func() error {
+			appCtx := new(v1alpha2.ApplicationContext)
+			err := k8sClient.Get(ctx, key, appCtx)
+			if err != nil {
+				return err
+			}
+			if len(appCtx.Status.Workloads) != 1 {
+				return fmt.Errorf("appContext status error:the number of workloads not right")
+			}
+			if appCtx.Status.Workloads[0].ComponentName != compName1 {
+				return fmt.Errorf("appContext status error:the name of component not right, expect %s", compName1)
+			}
+			cw := new(v1alpha2.ContainerizedWorkload)
+			return k8sClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: cwName1}, cw)
+		}, time.Second*60, time.Second*5).Should(BeNil())
```

Comment:
5 seconds as an interval seems to be too long, it will make this test run longer. I suggest using less than 1 second.

Reply:
Got it, will update